src/iarc7_abstract/roomba_request_executer.py

```diff
@@ -165,38 +236,64 @@ def _run(cls, roomba_request, status_callback):
                 cls._client.cancel_goal()
                 rospy.logwarn("Hold Position Task canceled")
 
-            # state transitioning
-            if (state != RoombaRequestExecuterState.HOLD_POSITION
-                    and (not cls._client.get_result()
-                        or not cls._client.get_result().success)):
-                if state == RoombaRequestExecuterState.RECOVER_FROM_FAILURE:
-                    state = RoombaRequestExecuterState.FAILED_TASK_AND_RECOVERY
-                elif state == RoombaRequestExecuterState.RECOVER_FROM_SUCCESS:
-                    state = RoombaRequestExecuterState.FAILED_RECOVERY
-                else:
-                    state = RoombaRequestExecuterState.RECOVER_FROM_FAILURE
-            else:
+            """state transitioning"""
+
+            # see http://docs.ros.org/api/actionlib_msgs/html/msg/GoalStatus.html
+            # for all status IDs
+            # 4: ABORTED
+            # 5: REJECTED
+            # 9: LOST
+            if cls._client.get_state() not in [4,5,9]:
+                # the task was successful
+
+                # Hit and Block roomba require recovery afterwards
                 if state == RoombaRequestExecuterState.HIT:
                     state = RoombaRequestExecuterState.RECOVER_FROM_SUCCESS
 
                 elif state == RoombaRequestExecuterState.BLOCK:
                     state = RoombaRequestExecuterState.RECOVER_FROM_SUCCESS
 
+                # transition after going to a roomba
+                elif state == RoombaRequestExecuterState.GOING_TO:
+                    if tracking_mode == RoombaRequest.BLOCK:
+                        state = RoombaRequestExecuterState.BLOCK
+                    elif tracking_mode == RoombaRequest.HIT:
+                        state = RoombaRequestExecuterState.HIT
+                    else:
+                        state = RoombaRequestExecuterState.FOLLOWING
+
+                # bump roomba logic
+                elif state == RoombaRequestExecuterState.FOLLOWING:
+                    if cls._roomba_turning:
+                        state = RoombaRequestExecuterState.BLOCK
```

Comment:
How is this state left? Is it supposed to be left?

Reply:
explained in person